boa/src/object/internal_methods/string.rs

```diff
@@ -0,0 +1,132 @@
+use crate::{
+    object::JsObject,
+    property::{PropertyDescriptor, PropertyKey},
+    Context, JsResult, JsValue,
+};
+
+use super::{InternalObjectMethods, ORDINARY_INTERNAL_METHODS};
+
+pub(crate) static STRING_EXOTIC_INTERNAL_METHODS: InternalObjectMethods = InternalObjectMethods {
+    __get_own_property__: string_exotic_get_own_property,
+    __define_own_property__: string_exotic_define_own_property,
+    __own_property_keys__: string_exotic_own_property_keys,
+    ..ORDINARY_INTERNAL_METHODS
+};
+
+/// Gets own property of 'String' exotic object
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-string-exotic-objects-getownproperty-p
+#[inline]
+pub(crate) fn string_exotic_get_own_property(
+    obj: &JsObject,
+    key: &PropertyKey,
+    context: &mut Context,
+) -> JsResult<Option<PropertyDescriptor>> {
+    let desc = super::ordinary_get_own_property(obj, key, context)?;
+
+    if desc.is_some() {
+        Ok(desc)
+    } else {
+        Ok(string_get_own_property(obj, key))
+    }
+}
+
+/// Defines own property of 'String' exotic object
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-string-exotic-objects-defineownproperty-p-desc
+#[inline]
+pub(crate) fn string_exotic_define_own_property(
+    obj: &JsObject,
+    key: PropertyKey,
+    desc: PropertyDescriptor,
+    context: &mut Context,
+) -> JsResult<bool> {
+    let string_desc = string_get_own_property(obj, &key);
+
+    if let Some(string_desc) = string_desc {
+        let extensible = obj.borrow().extensible;
+        Ok(super::is_compatible_property_descriptor(
+            extensible,
+            desc,
+            string_desc,
+        ))
+    } else {
+        super::ordinary_define_own_property(obj, key, desc, context)
+    }
+}
+
+/// Gets own property keys of 'String' exotic object
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-string-exotic-objects-ownpropertykeys
+#[inline]
+pub(crate) fn string_exotic_own_property_keys(
+    obj: &JsObject,
+    _context: &mut Context,
+) -> JsResult<Vec<PropertyKey>> {
+    let mut keys = Vec::new();
+
+    let obj = obj.borrow();
+
+    let string = obj
+        .as_string()
+        .expect("string exotic method should only be callable from string objects");
+    let len = string.encode_utf16().count();
+
+    for i in 0..len {
+        keys.push(i.into());
+    }
+
+    // todo: sort keys or ensure in some way that indexed properties are sorted... or maybe it's not necessary?
```

Comment:
I don't think the spec expects them to be sorted in any particular way, right?

Reply:
It does. In `[[OwnPropertyKeys]]`'s [spec](https://tc39.es/ecma262/#sec-string-exotic-objects-ownpropertykeys):
```
6. For each own property key P of O such that P is an array index and ! ToIntegerOrInfinity(P) ≥ len, in ascending numeric index order, do
    a. Add P as the last element of keys.
```

But I don't know if it's just for display purposes or if there are tests for this specific check.

Reply:
If it's in the spec we should implement it. I guess we can just use `sort_unstable()` in it.